### DISCUSSION TITLE
Serial I/O from mcsim to device connected over USB to host computer

### DISCUSSION THREAD

**Discussion by gingerbeardman** [Moddable-OpenSource/moddable#933] [Q&A] [answered]
## Background

[gingerbeardman](https://twitter.com/gingerbeardman/status/1563270523307831298): 
> Will mcsim, running as M5Paper, support serial i/o with a device connected over USB to my host PC? Please forgive my laziness on this question. Thanks!

[phoddie](https://twitter.com/phoddie/status/1563275353337999362): 
> No, not out-of-the-box. mcsim doesn't try to simulate hardware. But... there's a macOS implementation of ECMA-419 serial that should work. If you want to give it a try, I'm happy to help, but it could be a bit much for Twitter. ;) Just open a discussion on our GitHub repo.

## Goal
My goal is to use my M5Paper as an I/O device for my Playdate handheld console, using serial over USB. This kind of comms is already possible using a PC. 

I guess the first step is to create a small app to do serial I/O over USB from/to simulator to/from device?

Any pointers or help appreciated!

**Comment by phoddie** [comment, marked as answer]
The Moddable SDK is moving to using [ECMA-419](https://419.ecma-international.org) for I/O, which includes serial. The ECMA-419 APIs are well supported on ESP32 and ESP8266, so you are set there.

On macOS, there are implementations of ECMA-419 Serial, TCP, UDP, and Listener I/O. Because ESP32 and macOS are very different operating systems,, there are differences in configuration. The APIs, however, are identical.

I built a small exampleI starting from the [serial echo example](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/io/serial/echo/main.js). The modified version is configured for macOS and also outputs received bytes to the xsbug console. Here's the `main.js` script:

```js
import Serial from "embedded:io/serial";
import TextDecoder from "text/decoder";

const decoder = new TextDecoder;
new Serial({
	device: "/dev/cu.usbserial-0001",
	baud: 74880,
	format: "buffer",
	onReadable: function(count) {
		const buffer = this.read();
		this.write(buffer);
		trace(decoder.decode(buffer, {stream: true}), "\n");
	},
});
```

Here's the `manifest.json`:

```json
{
	"include": [
		"$(MODDABLE)/modules/io/manifest.json",
		"$(MODDABLE)/modules/data/text/decoder/manifest.json"
	],
	"modules": {
		"*": "./main"
	}
}
```

To test, I connected an ESP8266 to a USB port on my Mac (via a Moddable Programmer Serial-to-USB bridge) and hit the reset button to generate some bytes (the boot-loader's baud rate on ESP8266 is an unusual 74880). Adjust the `device` and `baud` as appropriate for your set-up.

**Reply by gingerbeardman** [reply]
Thanks, I'll try this out!

**Reply by gingerbeardman** [reply]
mcsim is giving me the following strange errors, these are from a single run through (after which it quits):

> /Users/matt/Library/Application Support/tech.moddable.xsbug/eval/15.js (84) # Break: MockupBehavior.prototype.onJSON: get model: undefined variable!
> /Users/matt/Library/Application Support/tech.moddable.xsbug/eval/31.js (84) # Break: MockupBehavior.prototype.onJSON: get model: undefined variable!
> /Users/matt/Library/Application Support/tech.moddable.xsbug/eval/47.js (84) # Break: MockupBehavior.prototype.onJSON: get model: undefined variable!
> /Users/matt/Library/Application Support/tech.moddable.xsbug/eval/47.js (84) # Break: MockupBehavior.prototype.onJSON: get model: undefined variable!
> /Users/matt/Library/Application Support/tech.moddable.xsbug/eval/63.js (84) # Break: MockupBehavior.prototype.onJSON: get model: undefined variable!
> /Users/matt/Library/Application Support/tech.moddable.xsbug/eval/63.js (84) # Break: MockupBehavior.prototype.onJSON: get model: undefined variable!

I'm using the following to open the port to Playdate and read what is being sent:

> device: "/dev/cu.usbmodemPDU1_Y0123456",	// Playdate
> baud: 115200,

usb demo app produces no output in xsbug console.

I am sure Playdate is outputting over serial because I have confirmed using
`picocom -b 115200 -p n -d 8 /dev/cu.usbmodemPDU1_Y0123456 --omap crcrlf`

I have run the Hello World example successfully.

Any ideas?

**Reply by gingerbeardman** [reply]
Strange! It runs fine if I do `mcconfig -d -m -p simulator` (though I have to manually switch to M5Paper layout)
but errors if I do `mcconfig -d -m -p simulator/m5paper`

Any ideas @FWeinb ?

**Reply by FWeinb** [reply]
The syntax to start the simulator in a specific layout is `mcconfig -d -m -p sim/m5paper`

**Reply by phoddie** [reply]
This is a funny one. Something recently changed in the load order of simulators (probably when the `Compartment` API was updated) that causes the M5Paper simulator to generate an exception when selected at start-up. You can easily work around that by editing `$MODDABLE/build/simulators/m5paper.js` to add this before line 84 where the exception occurs:

```
      if (!globalThis.model) return;
```

With the change in place, everything – including serial – works normally.

I don't have time today to track down the root cause, but we'll sort that out.

FWIW – you can use both `sim/m5paper` and `simulator/m5paper` for the `-p` option when launching `mcconfig`.

**Reply by phoddie** [reply]
Just to close this out, the M5Paper simulator is using the `model` global. mcsim did not intend for that global to be available to simulator implementations, but it has been until recently. When updating mcsim to the latest `Compartment` API, the `model` global was made inaccessible to simulators leading to this exception. `model` is only used to access the current behavior, which happens to be `this`, so the code can be simplified to:

```js
      const POWER_INFO_NAME = `POWER_INFO_${pin}`;
      if (POWER_INFO_NAME in this) {
        this[POWER_INFO_NAME].string = value ? "On" : "Off";
      }
```

This change will be incorporated into the next Moddable SDK, which will be live before the end of the week.

**Reply by gingerbeardman** [reply]
Thanks, that last one worked best for me. Onwards!
